CIP-0021/CIP-0021.md

```diff
@@ -0,0 +1,139 @@
+
+While the Cardano CDDL specification usually does not limit the size of integers, HW wallets only support `int64` for signed integers and `uint64` for unsigned integers. Any integer value must fit in the appropriate type.
+
+**Numbers of transaction elements**
+
+The numbers of inputs, outputs, certificates, withdrawals, tokens in an asset group, and other similar transaction elements must not exceed `UINT16_MAX`, i.e. 65535.
+
+**Optional empty lists and maps**
+
+Unless mentioned otherwise in this CIP, optional empty lists and maps must not be included as part of the transaction body or its elements.
+
+**Outputs**
+
+Outputs containing no multi-asset tokens must be serialized as a simple tuple, i.e. `[address, coin]` instead of `[address, [coin, {}]]`.
+
+**Multiassets**
+
+Since multiassets (`policy_id` and `asset_name`) are represented as maps, both need to be sorted in accordance with the specified canonical CBOR format. Also, an output or the mint field must not contain duplicate `policy_id`s and a policy must not contain duplicate `asset_name`s.
+
+**Certificates**
+
+Certificates of type `genesis_key_delegation` and `move_instantaneous_rewards_cert` are not supported and must not be included.
+
+If a transaction contains a pool registration certificate, then it must not contain:
+
+- any other certificate;
+- any withdrawal;
+- mint entry.
+
+It is allowed to arbitrarily combine other supported certificate types, but all the certificates included in a transaction must have the same type of stake credential i.e. either 0 - `addr_keyhash` or 1 - `scripthash`. The stake credential type must also be consistent with the type used for withdrawals.
+
+**Withdrawals**
+
+Since withdrawals are represented as a map of reward accounts, withdrawals also need to be sorted in accordance with the specified canonical CBOR format. A transaction must not contain duplicate withdrawals. All withdrawals included in a transaction must have the same type of stake credential i.e. either 0 - `addr_keyhash` or 1 - `scripthash`. The stake credential type must also be consistent with the type used for certificates.
+
+**Auxiliary data**
+
+HW wallets do not serialize auxiliary data because of their complex structure. They only include the given auxiliary data hash in the transaction body. The only exception is Catalyst voting registration because it requires a signature computed by the HW wallet.
+
+In this exceptional case, auxiliary data must be encoded in their "tuple" format:
+
+```
+[ transaction_metadata: { * transaction_metadatum_label => transaction_metadatum }, auxiliary_scripts: [ * native_script ]]
+```
+
+The `auxiliary_scripts` must be an array of length 0.
+
+### Transaction witnesses
+
+There are two limits on the number of witnesses:
+
+- an absolute limit of `UINT16_MAX`, i.e. 65535;
+- a relative limit dependent on the transaction body (essentially one witness per each input, each withdrawal and each certificate in a typical transaction).
```

Comment:
This second condition may be overly restrictive in Alonzo where it'll be allowed to request additional signatures which do not correspond to any input. This would be however specified by a dedicated field on transactions `required_signers` (key 14 in transaction body).

Reply:
This will change with the introduction of multi-sig to HW wallets which is currently being finalised. It will be possible to send `additional_witness_requests` to the HW wallets which will sign the transaction with a key derived from the path given in the request.